File: packages/octocode-minifier/src/core/apply.ts
import { minifyContentSync } from './minifier.js';
import {
  minifyJsonReadable,
  minifyCodeCore,
  minifyGeneralCore,
  minifyMarkdownCore,
  removeComments,
} from './strategies.js';
import type { FileTypeMinifyConfig } from '../types/index.js';
import { INDENTATION_SENSITIVE_NAMES, MINIFY_CONFIG } from '../types/index.js';
import { getExtension } from '../utils/fileExtension.js';

export function applyMinification(content: string, filePath: string): string {
  try {
    const minifiedContent = minifyContentSync(content, filePath);
    return minifiedContent.length < content.length ? minifiedContent : content;
  } catch {
    return content;
  }
}

const JSON_EXTS = new Set(['json', 'jsonc', 'json5']);

function getBaseName(filePath: string): string {
  return (filePath.split(/[\\/]/).pop() || '').toLowerCase();
}

function getStandardContentConfig(
  filePath: string,
  ext: string
): FileTypeMinifyConfig | undefined {
  const baseName = getBaseName(filePath);
  if (INDENTATION_SENSITIVE_NAMES.has(baseName)) {
    return { strategy: 'conservative', comments: 'hash' };
  }

  return MINIFY_CONFIG.fileTypes[ext];
}

function stripConfiguredComments(
  content: string,
  config: FileTypeMinifyConfig | undefined
): string {
  return config?.comments ? removeComments(content, config.comments) : content;
}

/**
 * Content-safe minification for file viewing tools (localGetFileContent,
 * githubGetFileContent, PR patches).
 *
 * Designed for AGENT READABILITY — preserves structure and indentation.
 *
 * Pipeline per file type:
 *   JSON/JSONC/JSON5 → minifyJsonReadable (strip JSONC comments/trailing commas
 *                       while preserving readable layout);
 *                       only returned if result is shorter (i.e. JSONC had comments)
 *   Markdown          → minifyMarkdownCore (HTML comments, quote-replies stripped)
 *   Code/config files → strip all configured comment syntaxes, then minifyCodeCore
 *                       (trailing whitespace + blank-line compression;
 *                        original indentation PRESERVED for agent readability)
 *   Plain text (txt/log/unknown) → minifyGeneralCore (whitespace + indent compression)
 *   Indentation-sensitive names → strip hash comments, then minifyCodeCore
 *
 * Always returns the original if the minified version is not shorter.
 */
export function applyContentViewMinification(
  content: string,
  filePath: string
): string {
  try {
    const ext = getExtension(filePath, { lowercase: true, fallback: 'txt' });
    const config = getStandardContentConfig(filePath, ext);

    let minified: string;

    if (JSON_EXTS.has(ext)) {
      // Readable JSON: strip JSONC noise, keep pretty-printing.
      // Falls back to original when nothing was stripped (already clean JSON).
      minified = minifyJsonReadable(content).content;
    } else if (config?.strategy === 'markdown') {
      minified = minifyMarkdownCore(content);
    } else {
      const stripped = stripConfiguredComments(content, config);
      // Code files: preserve indentation so agents keep structural context.
      // Plain-text files (txt/log/no registered type): allow indent compression.
      minified =
        !config || config.strategy === 'general'
          ? minifyGeneralCore(stripped)
          : minifyCodeCore(stripped);
    }

    return minified.length < content.length ? minified : content;
  } catch {
    return content;
  }
}

export { MINIFY_CONFIG };

// Skeleton (minify:"symbols") extraction lives in its own strategy-based module
// (AST for ts-js, lean heuristics for other families); re-exported here so
// both file-content tools keep their stable import path.
export {
  extractSignatures,
  SIGNATURES_ONLY_HINT,
} from '../signatures/extractSignatures.js';

File: packages/octocode-minifier-utils/src/apply.rs
use crate::config::{minify_config, indentation_sensitive_names};
use crate::file_extension::get_extension_internal;
use crate::minifier::minify_content_sync_inner;
use crate::strategies::{
    minify_json_readable_inner, minify_markdown_core,
    minify_general_core, minify_code_core, minify_js_oxc,
    minify_css_quality,
};
use crate::comment_remover::remove_comments;

/// Full minification — return minified if shorter, else original.
pub fn apply_minification_inner(content: &str, file_path: &str) -> String {
    let minified = std::panic::catch_unwind(|| minify_content_sync_inner(content, file_path))
        .unwrap_or_else(|_| content.to_owned());
    if minified.len() < content.len() { minified } else { content.to_owned() }
}

/// Content-view minification — agent-readable, preserves indentation.
/// Pipeline mirrors TS `applyContentViewMinification`.
pub fn apply_content_view_minification_inner(content: &str, file_path: &str) -> String {
    let result = std::panic::catch_unwind(|| {
        let ext = get_extension_internal(file_path, true, "txt");
        let basename = file_path
            .rsplit(['/', '\\'])
            .next()
            .unwrap_or(file_path)
            .to_lowercase();

        let cfg = if indentation_sensitive_names().contains(basename.as_str()) {
            minify_config().get("sh")
        } else {
            minify_config().get(ext.as_str())
        };

        if matches!(ext.as_str(), "json" | "jsonc" | "json5") {
            let (out, _) = minify_json_readable_inner(content);
            return out;
        }

        if cfg.map(|c| c.strategy) == Some("markdown") {
            return minify_markdown_core(content);
        }

        // P2: CSS / SCSS / LESS content-view: use lightningcss (much better than blank-line collapse)
        if matches!(ext.as_str(), "css"|"scss"|"less"|"sass") {
            return minify_css_quality(content);
        }

        // JS/TS: use OXC without mangling — preserves names for agent readability
        if matches!(ext.as_str(), "ts"|"tsx"|"js"|"jsx"|"mjs"|"cjs") {
            if let Some(oxc_out) = minify_js_oxc(content, file_path, false) {
                return oxc_out;
            }
            // OXC failed — fall through to comment-strip + code-core
        }

        let stripped = if let Some(c) = cfg {
            if let Some(groups) = c.comments {
                remove_comments(content, groups)
            } else {
                content.to_owned()
            }
        } else {
            content.to_owned()
        };

        match cfg.map(|c| c.strategy) {
            None | Some("general") => minify_general_core(&stripped),
            _                      => minify_code_core(&stripped),
        }
    }).unwrap_or_else(|_| content.to_owned());

    if result.len() < content.len() { result } else { content.to_owned() }
}
